ext/com_dotnet/tests/comtest/comtest.cpp

```diff
@@ -0,0 +1,318 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Christoph M. Becker  <cmb@php.net>                           |
+   | Based on: <https://thrysoee.dk/InsideCOM+/>                          |
+   +----------------------------------------------------------------------+
+ */
+
+#include "comtest.h" // Generated by MIDL
+
+long g_cComponents = 0;
+long g_cLocks = 0;
+
+class CDocument : public IDocument, IPersistStream
+{
+public:
+	// IUnknown
+	STDMETHODIMP_(ULONG) AddRef();
+	STDMETHODIMP_(ULONG) Release();
+	STDMETHODIMP QueryInterface(REFIID riid, void** ppv);
+
+	// IDispatch
+	STDMETHODIMP GetTypeInfoCount(UINT* pCountTypeInfo);
+	STDMETHODIMP GetTypeInfo(UINT iTypeInfo, LCID lcid, ITypeInfo** ppITypeInfo);
+	STDMETHODIMP GetIDsOfNames(REFIID riid, LPOLESTR* rgszNames, UINT cNames, LCID lcid, DISPID* rgDispId);
+	STDMETHODIMP Invoke(DISPID dispIdMember, REFIID riid, LCID lcid, WORD wFlags, DISPPARAMS* pDispParams,
+		VARIANT* pVarResult, EXCEPINFO* pExcepInfo, UINT* puArgErr);
+
+	// IDocument
+	STDMETHODIMP get_Content(BSTR* retvalue);
+	STDMETHODIMP put_Content(BSTR newvalue);
+
+	// IPersist
+	STDMETHODIMP GetClassID(CLSID *pClassID);
+
+	// IPersistStream
+	STDMETHODIMP IsDirty( void);
+	STDMETHODIMP Load(IStream *pStm);
+	STDMETHODIMP Save(IStream *pStm, BOOL fClearDirty);
+	STDMETHODIMP GetSizeMax(ULARGE_INTEGER *pcbSize);
+
+	CDocument() : m_content(SysAllocString(L"")), m_cRef(1) { g_cComponents++; }
+	~CDocument() { SysFreeString(m_content); g_cComponents--; }
+	HRESULT Init(void);
+
+private:
+	BSTR m_content;
+	ULONG m_cRef;
+	ITypeInfo* m_pTypeInfo;
+	BOOL m_dirty;
+};
+
+ULONG CDocument::AddRef()
+{
+	return ++m_cRef;
+}
+
+ULONG CDocument::Release()
+{
+	if (--m_cRef != 0) {
+		return m_cRef;
+	}
+	m_pTypeInfo->Release();
+	delete this;
+	return 0;
+}
+
+HRESULT CDocument::QueryInterface(REFIID riid, void** ppv)
+{
+	if (riid == IID_IUnknown) {
+		*ppv = (IDocument*) this;

```

Comment:
it might be preferable to use static_cast<> operator.

Reply:
Right. Makes sense.

Reply:
Looks ok, I did not say dynamic_cast on purpose here because, if I m not mistaken, I ve read somewhere these COM+ stuff does use RTTI so we cannot use this operator here, static_cast<> should do :)

Reply:
I think `static_cast` is exactly what we want, since it is guaranteed that these interfaces are actually implemented by `CDocument`.